### ISSUE TITLE
methods(part 2)

### ISSUE THREAD

**Issue by ddkorn** [ddkorn/diploma#2]
Объединение аминокислот в группы с общими физико-химическими свойствами
Данный подход предполагает, что замена аминокислоты на аминокислоту с близкими физико-химическими свойствами не влияет на функциональную активность клона.
Таким образом, аминокислоты в составе следующих групп считались идентичными: FYH, LIMV, TSC.
Классификаторы для предсказания COVID-статуса субъекта
Для прогнозирования статуса пациентов по набору клонотипов использовались модели машинного обучения Random Forest и SVM.  Оптимизация гиперпараметров моделей осуществлялась методом перекрёстной проверки (кросс-валидацией). Для наших целей использовалась кросс-валидация по 3 блокам с поиском оптимальных гиперпараметров  среди следующего о списка параметров.
Список параметров RF: {'criterion': ['gini'], 'max_depth' : [3, 5, 7], 'min_samples_leaf': [4, 10, 20], 'min_samples_split': [10, 20, 40], 'max_features': [15, 20, 50]}
Список параметров SVM: {"C": [0.01, 0.5, 1], "kernel": ["rbf"], "grade": [1, 2], "gamma": ["scale"]}
Поиск гиперпараметров выполнялся для наброра обучающих выборок, после чего  найденные гиперпараметры применялись к тестовой выборке.
При разделении пациентов на группы учитывались партии секвенирования, то есть не допускалось попадания в разные группы образцов принадлежащих одной партии секвенирования. Также мы стремились добиться количественного баланса между классами COVID+ и COVID- в обучающих и тестовых выборках. Таким образом  были сформированы следующие группы пациентов.
rus1:"2020 09 27", "2020 12 18", "2020 10 19"
rus2: "2020 09 05", "2020 09 28", "2021 01 07"
адаптивный1: «COVID-19-NIH/NIAID», «COVID-19-HUniv12Oct», «COVID-19-адаптивный», «KECK»
адаптивный2: "COVID-19-DLS", "COVID-19-ISB", "COVID-19-IRST/AUSL", "HIP"
Затем были получены комбинации (размещение) упомянутых групп и для каждой комбинации рассчитывались оценки точности моделей. Результат обобщен на рисунках ...

Была подсчитана частота встречающихся  комбинаций V и J генных сегментов для идентифицированных клонов. Посредством поиска наиболее распространенные комбинации генов Variable-Joining  можно выявить широко представленные  в анализируемой группе клонотипы. Зачастую смещение частот вариантов  V и J обусловлено не биологическими фактора, а условиями проведения эксперимента. В частности, использование разного микса праймеров в различных сериях постановки секвенирования может привести к значительному сдвига V,J –usage в соответствующих партиях образцов.

Нормировка  V-usage
Подсемейства V сегментов объединялись в семейства (например, TRBV6-1 и TRBV6-4  трансформировались в TRBV6). (2) Вычислялись V-usage для каждого V сегмента в каждом батче (Adaptive_acute, HIP, Keck, FMBA, VIP_7) и  затем медианное значение V-usage между батчами (median(TRBV6_in_FMBA,TRBV6_in_HIP, ...))  (3) Находился шкалирующий коэффициент для каждого V сегмента в каждом батче. Таким образом после домножения частот всех клонотипов с данным V сегментом в данном батче на шкалирующий	 коэффициент итоговый V-usage становился равным медианному  V-usage. (4)  Сумма  частот V-сегментов  в каждом образце приводилась к единице (Сумма  частот  делилась на сумму изменённых частот).

Отбор признаков для машинного обучения
Важным этапом машинного обучения являлся отбор признаков – последовательностей CDR3 и эффективного представления их в виде вектора значений. Строго говоря, перед нами стояла биологическая задача поиска клонотипов, а также паттернов объединяющих эти последователь6ности, с помощью которых можно было бы различать образцы здоровых доноров и переболевших COVID. 
Мы использовали две стратегии поиска значимых для модели CDR3. Первый подход основан на использование публичных клонов в качестве признаков для обучения.  Он обоснован предположением, что репертуары TКР у пациентов перенесших короновирусную инфекцию близки по сотаву и по представленности публичных клонов. С другой стороны образцы здоровых доноров также вероятно содержат общие специфичные клонотипы.  Таким  образом на вход классификатора подается множество  (~10000) наиболее представленных среди анализируемых образцов публичных клонов.
Вторая стратегия отбора включает в себя предварительный этап поиска ассоциированных с COVID последовательных среди пула идентифицированных образцов и выявления наиболее распространенных мотивов последовательностей  с помощью методов кластеризации. На вход модели подаются относительно небольшое число (~2000) клонов с заранее известной  специфичность к COVID.
Кроме того мы исплользовали два способа предстваления фичей в бразце. В первом случае учитывалась частота встретившегося клона в образце, такое предствавление называется взвешенное. В противоположном подходе вес клона не учитывается, оценивается только факт вхождения клонотипа в данный репертуар (0, если клонотип не найден, 1, если найден).  
Математическая модель 
Задачу классификации можно представить в виде математической абстракции:
X1, X2…Xn ~y,
Где  X i  представление данного признака в виде числового занчения.  Каждый образец описывается n-мерным вектор значений X. Каждому образцу присваивается метка yi, где число I соответствует числу классов.В данном  частном случае , решением задачи классификации является алгоритм  получающий на входе множество  Xii и способный  классифицировать данный произвольный образец  (присвоить статус COVID или healthy). Функция  L(x,y) определяется как функция потерь, которая оптимизируется для решения данной задачи. 
Мы использовали две модели машинного обучения: Random Forest и SVM. 
Random Forest

случайный лес строит несколько деревьев решений и объединяет их вместе, чтобы получить более точный и стабильный прогноз.

Случайный лес работает в два этапа: создается случайный лес путем объединения N деревьев решений, затем  рассчитывается прогноз для каждого дерева, созданного на первом этапе.

Алгоритм можно описать с помощью следующих шагов:
Шаг 1: Выбираем случайные точки данных K из тренировочного набора.
Шаг 2: Строим деревья решений выбранных точек данных.
Шаг 3: Комбинируем N деревьев решений.
Шаг 4: Повторяем шаги 1 и 2.
Шаг 5: Для новых точек данных получаем прогнозы каждого дерева решений и присваиваем точкам данных классы, которые набирают большинство голосов.
SVM
https://towardsdatascience.com/support-vector-machine-introduction-to-machine-learning-algorithms-934a444fca47
Цель алгоритма метода опорных векторов — найти гиперплоскость в N-мерном пространстве (N — количество признаков), которая наиболее эффективно разделяет точки данных.
Чтобы разделить два класса точек данных, можно выбрать множество возможных гиперплоскостей. Задача алгоритма найти плоскость с максимальным зазором -  максимальным расстоянием между точками данных обоих классов.
Опорные векторы — это точки данных, расположенные наиболее близко к гиперплоскости и влияющие на положение и ориентацию гиперплоскости. С помощью опорных векторов, максимизируется зазор классификатора. Удаление опорных векторов изменяет положение гиперплоскости. 
In the SVM algorithm, we are looking to maximize the margin between the data points and the hyperplane. The loss function that helps maximize the margin is hinge loss.
Hinge loss function (function on left can be represented as a function on the right)
The cost is 0 if the predicted value and the actual value are of the same sign. If they are not, we then calculate the loss value. We also add a regularization parameter the cost function. The objective of the regularization parameter is to balance the margin maximization and loss. After adding the regularization parameter, the cost functions looks as below.
Loss function for SVM
Now that we have the loss function, we take partial derivatives with respect to the weights to find the gradients. Using the gradients, we can update our weights.
Gradients
When there is no misclassification, i.e our model correctly predicts the class of our data point, we only have to update the gradient from the regularization parameter.
Обучающие, контрольные и валидационные наборы данных
Чтобы обучить нашу модель и оценить ее качество, необходимо разделить совокупность данных на 3 части:
для непосредственного обучения модели обучающая выборка (training sample),  по которой производится настройка (оптимизация гиперпараметров) алгоритма
для оценки качества модели используется тестовая (контрольная) выборка (test sample), которая, не должна зависеть от обучающей;
для выбора наилучшей модели машинного обучения понадобится проверочная (валидационная) выборка (validation sample), которая также не должна пересекаться с обучающей выборкой. 

Метрики качества модели
Точность – доля правильных ответов. Однако этот показатель очень чувствителен к балансу размеров классов, поэтому его следует использовать с осторожностью.
Сбалансированная точность – среднее значение точности для каждого класса. Эта метрика более надежна в случае несбалансированного размера классов.
AUC-ROC – площадь под ROC-кривой, где ROC-кривая – это графическая диаграмма, иллюстрирующая диагностическую способность системы бинарного классификатора при изменении ее порога различения. ROC-кривая строится как истинная положительная частота по сравнению с ложноположительной, которая, в свою очередь, зависит от точности и полноты. Чтобы найти наилучший компромисс между точностью и полнотой модели, используется мера F1.
Precision  (точность )  доля объектов положительного класса, которые  являются истинно положительными
Recall (полнота) доля объектов положительного класса из всех объектов положительного класса, которач  была выявлена моделью.
https://habr.com/ru/company/ods/blog/328372/
